cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/RemoteSourceControlOperationRunner.java

```diff
@@ -0,0 +1,120 @@
+                                     RemoteClientFactory remoteClientFactory) {
+    int readTimeout = cConf.getInt(Constants.TaskWorker.SOURCE_CONTROL_HTTP_CLIENT_READ_TIMEOUT_MS);
+    int connectTimeout = cConf.getInt(Constants.TaskWorker.SOURCE_CONTROL_HTTP_CLIENT_CONNECTION_TIMEOUT_MS);
+    HttpRequestConfig httpRequestConfig = new HttpRequestConfig(connectTimeout, readTimeout, false);
+    this.remoteTaskExecutor = new RemoteTaskExecutor(cConf, metricsCollectionService, remoteClientFactory,
+                                                     RemoteTaskExecutor.Type.TASK_WORKER, httpRequestConfig);
+  }
+
+  @Override
+  public PushAppResponse push(PushAppContext pushAppContext)
+    throws PushFailureException, NoChangesToPushException, AuthenticationConfigException {
+    try {
+      RunnableTaskRequest request =
+        RunnableTaskRequest.getBuilder(PushAppTask.class.getName()).withParam(GSON.toJson(pushAppContext)).build();
+
+      LOG.debug("Pushing application {} to linked repository", pushAppContext.getAppToPush());
+      byte[] result = remoteTaskExecutor.runTask(request);
+      return GSON.fromJson(new String(result, StandardCharsets.UTF_8), PushAppResponse.class);
+    } catch (RemoteExecutionException e) {
+      RemoteTaskException remoteTaskException = e.getCause();
+      String exceptionClass = remoteTaskException.getRemoteExceptionClassName();
+      String exceptionMessage = remoteTaskException.getMessage();
+
+      if (NoChangesToPushException.class.getName().equals(exceptionClass)) {
+        throw new NoChangesToPushException(exceptionMessage);
+      }
+
+      if (AuthenticationConfigException.class.getName().equals(exceptionClass)) {
+        throw new AuthenticationConfigException(exceptionMessage, remoteTaskException.getCause());
+      }
+
+      throw new PushFailureException(exceptionMessage, remoteTaskException.getCause());
+    } catch (Exception ex) {
+      throw new PushFailureException(ex.getMessage(), ex.getCause());
+    }
+  }
+
+  @Override
+  public ListAppsResponse list(NamespaceId namespace, RepositoryConfig repoConfig) throws ListAppsFailureException,
+    AuthenticationConfigException {
+    try {
+      ListAppsContext context = new ListAppsContext(namespace, repoConfig);
+      RunnableTaskRequest request =
+        RunnableTaskRequest.getBuilder(ListApplicationsTask.class.getName()).withParam(GSON.toJson(context)).build();
+
+      LOG.debug("Listing applications in linked repository for namespace {}", namespace);
+      byte[] result = remoteTaskExecutor.runTask(request);
+      return GSON.fromJson(new String(result, StandardCharsets.UTF_8), ListAppsResponse.class);
+    } catch (RemoteExecutionException e) {
+      RemoteTaskException remoteTaskException = e.getCause();
+      Throwables.propagateIfInstanceOf(remoteTaskException.getCause(), AuthenticationConfigException.class);
```

Comment:
Currently because of how `RemoteTaskException` is implemented, we cannot directly check the exception instance type this way. Instead we need to check the className like how we did in the above push flow

Reply:
Changed